server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java

```diff
@@ -156,86 +162,98 @@ private void flushToFile(ShuffleDataFlushEvent event) {
-            ShuffleServerMetrics.counterWriteTotal.inc();
-            if (writeTime > writeSlowThreshold) {
-              ShuffleServerMetrics.counterWriteSlow.inc();
-            }
-            updateCommittedBlockIds(event.getAppId(), event.getShuffleId(), blocks);
-            writeSuccess = true;
-            ShuffleServerMetrics.counterTotalWriteDataSize.inc(event.getSize());
-            ShuffleServerMetrics.counterTotalWriteBlockSize.inc(event.getShuffleBlocks().size());
-            if (event.getSize() < eventSizeThresholdL1) {
-              ShuffleServerMetrics.counterEventSizeThresholdLevel1.inc();
-            } else if (event.getSize() < eventSizeThresholdL2) {
-              ShuffleServerMetrics.counterEventSizeThresholdLevel2.inc();
-            } else if (event.getSize() < eventSizeThresholdL3) {
-              ShuffleServerMetrics.counterEventSizeThresholdLevel3.inc();
-            } else {
-              ShuffleServerMetrics.counterEventSizeThresholdLevel4.inc();
-            }
-          } catch (Exception e) {
-            LOG.warn("Exception happened when write data for " + event + ", try again", e);
-            ShuffleServerMetrics.counterWriteException.inc();
-            Thread.sleep(1000);
-          } finally {
-            if (useMultiStorage) {
-              lock.readLock().unlock();
-            }
-          }
-          retry++;
+          writeSuccess = true;
+          break;
+        } catch (Exception e) {
+          LOG.warn("Exception happened when write data for " + event + ", try again", e);
+          ShuffleServerMetrics.counterWriteException.inc();
+          Thread.sleep(1000);
         }
+
+        retry++;
       }
     } catch (Exception e) {
       // just log the error, don't throw the exception and stop the flush thread
       LOG.error("Exception happened when process flush shuffle data for " + event, e);
     } finally {
-      if (shuffleServer != null) {
-        shuffleServer.getShuffleBufferManager().releaseMemory(event.getSize(), true, false);
-        long duration = System.currentTimeMillis() - start;
-        if (writeSuccess) {
-          LOG.debug("Flush to file success in " + duration + " ms and release " + event.getSize() + " bytes");
-        } else {
-          LOG.error("Flush to file for " + event + " failed in "
-              + duration + " ms and release " + event.getSize() + " bytes");
+      long duration = System.currentTimeMillis() - start;
+      processWriteResult(event, duration, writeSuccess);
+    }
+  }
+
+  private void processWriteResult(ShuffleDataFlushEvent event, long duration, boolean writeSuccess) {
+    if (shuffleServer != null) {
+      shuffleServer.getShuffleBufferManager().releaseMemory(event.getSize(), true, false);
+      storageErrorMap.putIfAbsent(event.getStorageIndex(), new AtomicInteger(0));
+      AtomicInteger errors = storageErrorMap.get(event.getStorageIndex());
+      if (writeSuccess) {
+        errors.updateAndGet(x -> {
+          --x;
```

Comment:
`errors.updateAndGet(x -> --x < 0 ? 0 : x)`?

Reply:
I prefer the current implement. Because it's more easy to understand.